source/material_model/diffusion_dislocation.cc

```diff
@@ -173,8 +177,60 @@ namespace aspect
                                   prefactor_stress_dislocation *
                                   std::pow(stress_ii, stress_exponents_dislocation[j]-1);
 
-              stress_ii -= strain_rate_residual/strain_rate_deriv;
+              // If the strain rate derivative is zero, we catch it below.
+              if (strain_rate_deriv>std::numeric_limits<double>::min())
+                stress_ii -= strain_rate_residual/strain_rate_deriv;
               stress_iteration += 1;
+
+              // In case the Newton iteration does not succeed, we do a fixpoint iteration to see
+              // where the problem is and what the viscosities should be.
+              // Check if anything that would be used in the next iteration is not finite.
+              const bool abort_newton_iteration = !numbers::is_finite(stress_ii)
+                                                  || !numbers::is_finite(strain_rate_residual)
+                                                  || !numbers::is_finite(strain_rate_deriv)
+                                                  || strain_rate_deriv < std::numeric_limits<double>::min()
+                                                  || !numbers::is_finite(std::pow(stress_ii, stress_exponents_diffusion[j]-1))
+                                                  || !numbers::is_finite(std::pow(stress_ii, stress_exponents_dislocation[j]-1))
+                                                  || stress_iteration == stress_max_iteration_number;
+              if (abort_newton_iteration)
+                {
+                  double diffusion_strain_rate = edot_ii;
+                  double dislocation_strain_rate = min_strain_rate;
+                  stress_iteration = 0;
+
+                  do
+                    {
+                      const double old_diffusion_strain_rate = diffusion_strain_rate;
+
+                      const double diffusion_viscosity = std::min(std::max(0.5 * std::pow(prefactors_diffusion[j],-1.0/stress_exponents_diffusion[j])
```

Comment:
I tried, but I am not sure it is easier to read now. The original version was the formula exactly as stated in the description of this material model at the bottom of this file.